giraffe/src/transforms/mosaic.ts

```diff
@@ -1,121 +1,180 @@
+    yColumnKeys.forEach(columnKey => {
+      const column = inputTable.getColumn(columnKey, 'string')
+      yInputCols[columnKey] = column
+    })
+  }
+
+  // Mosaic can only have one column as the fill value,
+  //   always the first fill column key
   const valueKey = fillColumnMap.columnKeys[0]
 
+  const timeStampMap = {}
+
+  for (let i = 0; i < inputTable.length; i++) {
+    const yColumnTick = Array.isArray(yColumnKeys)
+      ? yColumnKeys.reduce((combinedValue, key) => {
+          let value = ''
+          if (yInputCols[key]) {
+            value = yInputCols[key][i]
+          }
+          return combinedValue + value
+        }, '')
+      : ''
+
+    const yTickLabel = labelColumns.reduce((combinedValue, key) => {
+      let value = ''
+      if (yInputCols[key]) {
+        value = yInputCols[key][i]
+      }
+      return combinedValue
+        ? `${combinedValue}${yLabelColumnSeparator}${value}`
+        : value
+    }, '')
+
+    const currentX = xInputCol[i]
+    const currentFillValue = fillColumnMap.mappings[fillColumn[i]][valueKey]
+
+    if (!timeStampMap[currentX]) {
+      timeStampMap[currentX] = []
+    }
+    timeStampMap[currentX].push({
+      yTickLabel,
+      yColumnTick,
+      fill: currentFillValue,
+    })
+  }
+
+  const sortedTimeStamps = Object.keys(timeStampMap).sort()
   let tableLength = 0
+
   /*
     This is the structure used to create and update the data map below:
 
-    dataMap[series] = [
-      prevValue: string,
+    dataMap[series] = {
       xMin: NumericColumnData,
       xMax: NumericColumnData,
-      values: NumericColumnData,
-      series: string[]
-    ]
+      fill: NumericColumnData,
+      series: string[],
+      displayedColumns: string[],
+      yTickLabel: string,
+    }
   */
   const dataMap = {}
 
-  for (let i = 0; i < inputTable.length; i++) {
-    const isOverlappingTimestamp =
-      xInputCol[i - 1] > xInputCol[i] && dataMap[yInputCol[i]]
-    if (isOverlappingTimestamp) {
-      break
-    }
-    const seriesExistsInDataMap = yInputCol[i] in dataMap
-    if (!seriesExistsInDataMap) {
-      //create series entry in dataMap
-      dataMap[yInputCol[i]] = [
-        fillColumnMap.mappings[fillColumn[i]][valueKey],
-        [xInputCol[i]],
-        [],
-        [fillColumnMap.mappings[fillColumn[i]][valueKey]],
-        [],
-      ]
-      continue
-    }
-    const valueIntervalEnds =
-      fillColumnMap.mappings[fillColumn[i]][valueKey] !=
-      dataMap[yInputCol[i]][0]
-    if (valueIntervalEnds) {
-      //update series entry in dataMap
-      dataMap[yInputCol[i]][0] = fillColumnMap.mappings[fillColumn[i]][valueKey]
-      dataMap[yInputCol[i]][1].push(xInputCol[i])
-      dataMap[yInputCol[i]][2].push(xInputCol[i])
-      dataMap[yInputCol[i]][3].push(
-        fillColumnMap.mappings[fillColumn[i]][valueKey]
-      )
-      dataMap[yInputCol[i]][4].push(yInputCol[i])
-
-      tableLength += 1
-    }
-  }
+  sortedTimeStamps.forEach(timeStamp => {
```

Comment:
It does need to exist because the parsed Flux results are not guaranteed to be in time order. Mosaic has to be sorted in time order because a mosaic "piece" is a time range of a particular value or status.

Reply:
would you mind adding a quick comment saying time order matters for mosaic but isn't guaranteed by flux?